pkg/controller/elasticsearch/services/services.go

```diff
@@ -181,26 +183,59 @@ func getServiceByName(c k8s.Client, es esv1.Elasticsearch, serviceName string) (
 	return svc, nil
 }
 
-// ElasticsearchURL calculates the base url for Elasticsearch, taking into account the currently running pods.
-// If there is an HTTP scheme mismatch between spec and pods we switch to requesting individual pods directly
-// otherwise this delegates to ExternalServiceURL.
-func ElasticsearchURL(es esv1.Elasticsearch, pods []corev1.Pod) string {
-	var schemeChange bool
-	for _, p := range pods {
-		scheme, exists := p.Labels[label.HTTPSchemeLabelName]
-		if exists && scheme != es.Spec.HTTP.Protocol() {
-			// scheme in existing pods does not match scheme in spec, user toggled HTTP(S)
-			schemeChange = true
+type urlProvider struct {
+	pods   func() []corev1.Pod
```

Comment:
Is there any reason not to return early in case of an error? 
```suggestion
	pods   func() ([]corev1.Pod, error)
```